File: two_sum.java
// 两数之和
// https://leetcode-cn.com/problems/two-sum/description/

class Solution {
    public int[] twoSum(int[] nums, int target) {
        for (int i = 0; i < nums.length; ++i) {
            for (int j = i + 1; j < nums.length; ++j) {
                if (target == (nums[i] + nums[j])) {
                    int[] retval = new int[2];
                    retval[0] = i;
                    retval[1] = j;
                    return retval;
                }
            }
        }
        return null;
    }
}

File: two_sum.ex
# 两数之和
# https://leetcode-cn.com/problems/two-sum/description/

defmodule Solution do
  @spec two_sum(nums :: [integer], target :: integer) :: [integer]
  def two_sum(nums, target) do
    two_sum(%{}, nums, 0, target)
  end

  defp two_sum(_, [], _, _) do
    [-1, -1]
  end

  defp two_sum(hashes, [head | tail], index, target) do
    if Map.has_key?(hashes, head) do
      [Map.get(hashes, head), index]
    else
      two_sum(Map.put(hashes, target - head, index), tail, index + 1, target)
    end
  end
end
